### PR TITLE
Separate norm and scale computation

### PR DESCRIPTION
**Pull request by Giuseppe5**

### CODE REVIEW

**Comment by Giuseppe5** [comment]
Tests are going to fail due to API changes. Let me know if I need to update also tests accordingly, and if the PR should be opened on a different "feature" branch.

File: brevitas/core/norm.py [hunk 1]
```diff
@@ -0,0 +1,105 @@
...
+
+
+class NormImplType(AutoName):
+    SAME_AS_SCALING = auto()
+    MAX = auto()
+    MAX_AVE = auto()
+    MAX_L2 = auto()
+
+
+class MaxParameterListNorm(torch.jit.ScriptModule):
```
Comment (volcacius) [inline brevitas/core/norm.py:65]: Remove Max from name since in the feature we might support other strategies too

File: brevitas/core/scaling.py [hunk 2]
```diff
@@ -61,6 +61,7 @@ class ScalingImplType(AutoName):
     AFFINE_STATS = auto()
     PARAMETER = auto()
     PARAMETER_FROM_STATS = auto()
+    BUFFER_FROM_STATS = auto()
```
Comment (volcacius) [inline brevitas/core/scaling.py:64]: Remove, this is gonna be a separate future PR

File: brevitas/core/scaling.py [hunk 3]
```diff
@@ -120,6 +121,39 @@ def _load_from_state_dict(self, state_dict, prefix, local_metadata, strict,
             missing_keys.remove(value_key)
 
 
+class BufferScaling(torch.jit.ScriptModule):
```
Comment (volcacius) [inline brevitas/core/scaling.py:124]: Remove

File: brevitas/core/stats.py [hunk 4]
```diff
@@ -68,8 +68,10 @@ class StatsOp(AutoName):
     MAX = auto()
     AVE = auto()
     MAX_AVE = auto()
+    MAX_L2 = auto()
     MEAN_SIGMA_STD = auto()
     MEAN_LEARN_SIGMA_STD = auto()
+    SAT_MAX_L2 = auto()
```
Comment (volcacius) [inline brevitas/core/stats.py:74]: Remove, this is separate work that can go in a separate future PR

File: brevitas/core/stats.py [hunk 5]
```diff
@@ -124,6 +126,39 @@ def state_dict(self, destination=None, prefix='', keep_vars=False):
...
+
+    @torch.jit.script_method
+    def forward(self, x: torch.Tensor):
+        per_channel_max = torch.max(torch.abs(x), dim=self.reduce_dim)[0]
+        out = torch.norm(per_channel_max, p=2)
+        out = out / math.sqrt(per_channel_max.view(-1).shape[0])
+        return out
+
+
+class SatMaxL2(torch.jit.ScriptModule):
```
Comment (volcacius) [inline brevitas/core/stats.py:144]: Remove

File: brevitas/core/stats.py [hunk 6]
```diff
@@ -296,6 +331,136 @@ def _load_from_state_dict(self, state_dict, prefix, local_metadata, strict,
...
+            missing_keys, unexpected_keys, error_msgs)
+        running_stats_key = prefix + 'running_stats'
+        if config.IGNORE_MISSING_KEYS and running_stats_key in missing_keys:
+            missing_keys.remove(running_stats_key)
+        training_key = prefix + 'training' # Pytorch stores training flag as a buffer with JIT enabled
+        if training_key in missing_keys:
+            missing_keys.remove(training_key)
+
+
+class RuntimeRestats(torch.jit.ScriptModule):
```
Comment (volcacius) [inline brevitas/core/stats.py:384]: Remove, this is gonna go in a separate future PR

File: brevitas/nn/quant_conv.py [hunk 7]
```diff
@@ -151,13 +152,16 @@ def __init__(self,
                 weight_scaling_shape = SCALING_SCALAR_SHAPE
                 weight_scaling_stats_reduce_dim = None
 
-            if weight_scaling_stats_op == StatsOp.MAX_AVE:
+            if weight_scaling_stats_op == StatsOp.MAX_AVE or \
+                    weight_scaling_stats_op == StatsOp.MAX_L2 or \
+                    weight_scaling_stats_op == StatsOp.SAT_MAX_L2:
```
Comment (volcacius) [inline brevitas/nn/quant_conv.py:157]: Remove

File: brevitas/proxy/parameter_quant.py [hunk 8]
```diff
@@ -153,6 +156,13 @@ def _weight_quant_init_impl(bit_width: Optional[int],
                                                  restrict_scaling_type=restrict_scaling_type,
                                                  is_parameter=True,
                                                  scaling_min_val=scaling_min_val)
+            elif scaling_impl_type == ScalingImplType.BUFFER_FROM_STATS:
```
Comment (volcacius) [inline brevitas/proxy/parameter_quant.py:159]: Remove

File: brevitas/proxy/parameter_quant.py [hunk 9]
```diff
@@ -133,7 +135,8 @@ def _weight_quant_init_impl(bit_width: Optional[int],
 
         elif scaling_impl_type == ScalingImplType.STATS \
                 or scaling_impl_type == ScalingImplType.AFFINE_STATS \
-                or scaling_impl_type == ScalingImplType.PARAMETER_FROM_STATS:
+                or scaling_impl_type == ScalingImplType.PARAMETER_FROM_STATS \
+                or scaling_impl_type == ScalingImplType.BUFFER_FROM_STATS:
```
Comment (volcacius) [inline brevitas/proxy/parameter_quant.py:139]: Remove

File: brevitas/proxy/parameter_quant.py [hunk 10]
```diff
@@ -200,13 +210,26 @@ def _weight_quant_init_impl(bit_width: Optional[int],
...
 
-            if bit_width_impl_type == BitWidthImplType.PARAMETER or \
+            if bit_width_impl_type != BitWidthImplType.CONST or \
                     bit_width_impl_type == BitWidthImplType.CONST and \
-                    scaling_impl_type == ScalingImplType.PARAMETER_FROM_STATS:
+                    norm_impl_type == NormImplType.SAME_AS_SCALING and \
+                    scaling_impl_type != ScalingImplType.STATS:
                 tensor_clamp_impl = TensorClamp()
             else:
                 tensor_clamp_impl = TensorClampSte()
```
Comment (volcacius) [inline brevitas/proxy/parameter_quant.py:219]: I need to rethink whether this is shifting logic behind what is currently being added

File: brevitas/proxy/runtime_quant.py [hunk 11]
```diff
@@ -201,6 +202,7 @@ def __init__(self,
                                               signed=signed,
                                               restrict_scaling_type=restrict_scaling_type)
                 tensor_quant = RescalingIntQuant(signed=signed,
```
Comment (volcacius) [inline brevitas/proxy/runtime_quant.py:204]: This goes away

Comment (Giuseppe5) [reply brevitas/proxy/runtime_quant.py:204]: How can I remove this, considering the changes [here?](https://github.com/Xilinx/brevitas/pull/128/files/2b72ab0e47a20c8d619d1e0011f4da8068aff64c#diff-479964b2d3ffefe1d9bb8a0d30482df6R493)

Comment (volcacius) [reply brevitas/proxy/runtime_quant.py:204]: You are right sorry, I was thinking about Restats

**Comment by volcacius** [comment]
Closing, pushing this branch as a feature branch within Xilinx/brevitas.

**Pull request closed without merging**